design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,312 @@
+
+Any changes to the CRD must be future safe and cleanly allow extension
+to further technologies (such as LDAP authentication).
+
+### Non-Goals
+
+* Support for Keystone API versions below v3. API version v2 has long
+  been deprecated and [has been removed in the "queens" version of
+  Keystone](https://docs.openstack.org/keystone/xena/contributor/http-api.html)
+  which was released in 2018 and is now in extended maintenance mode
+  (which means it gets no more points releases and only sporadic bug
+  fixes/security fixes).
+
+* Authenticating Ceph RGW to Keystone via admin token – Only
+  authentication via an OpenStack service account will be supported.
+  This is a deliberate choice as [admin tokens should not be used in production environments](
+  https://docs.openstack.org/keystone/rocky/admin/identity-bootstrap.html#using-a-shared-secret).
+
+* Support for APIs beside S3 and Swift.
+
+* Interaction of Swift with OBCs is out of scope of this document.  If
+  you need to access a bucket created by OBC via Swift you need to
+  create a separate `cephobjectstoreuser`, configure its access rights
+  to the bucket and use those credentials.
+
+* Support for authentication technologies other than Keystone (e.g. LDAP)
+
+* Exposing options that disable security features (e.g. TLS verification)
+
+## Proposal details
+
+The Object Store CRD will have to be extended to accommodate the new
+settings.
+
+### Keystone integration
+
+A new optional section `auth.keystone` is added to the Object Store
+CRD to configure the keystone integration:
+
+```yaml
+auth:
+  keystone:
+    url: https://keystone:5000/                     [*, r]
+    acceptedRoles: ["_member_", "service", "admin"] [*, r]
+    apiVersion: 3                                   [1, *]
+    implicitTenants: swift                          [*]
+    tokenCacheSize: 1000                            [*]
+    revocationInterval: 1200                        [*]
+    serviceUserSecretName: rgw-service-user         [2, r]
+```
+Annotations:
+* `[1]` Only support for Keystone API version 3 is specified in this
+  design document. Although RGW supports it, Keystone API v2 is
+  deliberately left out.
+* `[2]` The name of the secret containing the credentials for the
+  service user account used by RGW. It has to be in the same namespace
+  as the object store resource.
+* `[*]` These options map directly to [RGW configuration
+  options](https://docs.ceph.com/en/octopus/radosgw/config-ref/#keystone-settings),
+  the corresponding RGW option is formed by prefixing it with
+  `rgw_keystone_` and replacing upper case letters by their lower case
+  letter followed by an underscore. E.g. `tokenCacheSize` maps to
+  `rgw_keystone_token_cache_size`.
+* `[r]` These settings are required in the `keystone` section if
+  present.
+
+The certificate to verify the Keystone endpoint can't be explicitly
+configured in Ceph RGW, the system configuration of the pod is used,
```

Comment:
I'm not clear on this section. Maybe a new sentence here is valid? 
```suggestion
configured in Ceph RGW, the system configuration of the pod is used.
```

Reply:
I've reformulated this to be more readable and precise.